File: packages/firebase_admob/android/src/main/java/io/flutter/plugins/firebaseadmob/FirebaseAdMobPlugin.java
// Copyright 2017 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package io.flutter.plugins.firebaseadmob;

import android.app.Activity;
import android.view.Gravity;
import com.google.android.gms.ads.AdSize;
import com.google.android.gms.ads.MobileAds;
import com.google.firebase.FirebaseApp;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;
import java.util.Locale;
import java.util.Map;

public class FirebaseAdMobPlugin implements MethodCallHandler {

  private final Registrar registrar;
  private final MethodChannel channel;

  RewardedVideoAdWrapper rewardedWrapper;

  public static void registerWith(Registrar registrar) {
    if (registrar.activity() == null) {
      // If a background Flutter view tries to register the plugin, there will be no activity from the registrar.
      // We stop the registering process immediately because the firebase_admob requires an activity.
      return;
    }
    final MethodChannel channel =
        new MethodChannel(registrar.messenger(), "plugins.flutter.io/firebase_admob");
    channel.setMethodCallHandler(new FirebaseAdMobPlugin(registrar, channel));
  }

  private FirebaseAdMobPlugin(Registrar registrar, MethodChannel channel) {
    this.registrar = registrar;
    this.channel = channel;
    FirebaseApp.initializeApp(registrar.context());
    rewardedWrapper = new RewardedVideoAdWrapper(registrar.activity(), channel);
  }

  private void callInitialize(MethodCall call, Result result) {
    String appId = call.argument("appId");
    if (appId == null || appId.isEmpty()) {
      result.error("no_app_id", "a null or empty AdMob appId was provided", null);
      return;
    }
    MobileAds.initialize(registrar.context(), appId);
    result.success(Boolean.TRUE);
  }

  private void callLoadBannerAd(Integer id, Activity activity, MethodCall call, Result result) {
    String adUnitId = call.argument("adUnitId");
    if (adUnitId == null || adUnitId.isEmpty()) {
      result.error("no_unit_id", "a null or empty adUnitId was provided for ad id=" + id, null);
      return;
    }

    final Integer width = call.argument("width");
    final Integer height = call.argument("height");
    final String adSizeType = call.argument("adSizeType");

    if (!"AdSizeType.WidthAndHeight".equals(adSizeType)
        && !"AdSizeType.SmartBanner".equals(adSizeType)) {
      String errMsg =
          String.format(
              Locale.ENGLISH,
              "an invalid adSizeType (%s) was provided for banner id=%d",
              adSizeType,
              id);
      result.error("invalid_adsizetype", errMsg, null);
    }

    if ("AdSizeType.WidthAndHeight".equals(adSizeType) && (width <= 0 || height <= 0)) {
      String errMsg =
          String.format(
              Locale.ENGLISH,
              "an invalid AdSize (%d, %d) was provided for banner id=%d",
              width,
              height,
              id);
      result.error("invalid_adsize", errMsg, null);
    }

    AdSize adSize;
    if ("AdSizeType.SmartBanner".equals(adSizeType)) {
      adSize = AdSize.SMART_BANNER;
    } else {
      adSize = new AdSize(width, height);
    }

    MobileAd.Banner banner = MobileAd.createBanner(id, adSize, activity, channel);

    if (banner.status != MobileAd.Status.CREATED) {
      if (banner.status == MobileAd.Status.FAILED)
        result.error("load_failed_ad", "cannot reload a failed ad, id=" + id, null);
      else result.success(Boolean.TRUE); // The ad was already loaded.
      return;
    }

    Map<String, Object> targetingInfo = call.argument("targetingInfo");
    banner.load(adUnitId, targetingInfo);
    result.success(Boolean.TRUE);
  }

  private void callLoadInterstitialAd(MobileAd ad, MethodCall call, Result result) {
    if (ad.status != MobileAd.Status.CREATED) {
      if (ad.status == MobileAd.Status.FAILED)
        result.error("load_failed_ad", "cannot reload a failed ad, id=" + ad.id, null);
      else result.success(Boolean.TRUE); // The ad was already loaded.
      return;
    }

    String adUnitId = call.argument("adUnitId");
    if (adUnitId == null || adUnitId.isEmpty()) {
      result.error(
          "no_adunit_id", "a null or empty adUnitId was provided for ad id=" + ad.id, null);
      return;
    }
    Map<String, Object> targetingInfo = call.argument("targetingInfo");
    ad.load(adUnitId, targetingInfo);
    result.success(Boolean.TRUE);
  }

  private void callLoadRewardedVideoAd(MethodCall call, Result result) {
    if (rewardedWrapper.getStatus() != RewardedVideoAdWrapper.Status.CREATED
        && rewardedWrapper.getStatus() != RewardedVideoAdWrapper.Status.FAILED) {
      result.success(Boolean.TRUE); // The ad was already loading or loaded.
      return;
    }

    String adUnitId = call.argument("adUnitId");
    if (adUnitId == null || adUnitId.isEmpty()) {
      result.error(
          "no_ad_unit_id", "a non-empty adUnitId was not provided for rewarded video", null);
      return;
    }

    Map<String, Object> targetingInfo = call.argument("targetingInfo");
    if (targetingInfo == null) {
      result.error(
          "no_targeting_info", "a null targetingInfo object was provided for rewarded video", null);
      return;
    }

    rewardedWrapper.load(adUnitId, targetingInfo);
    result.success(Boolean.TRUE);
  }

  private void callShowAd(Integer id, MethodCall call, Result result) {
    MobileAd ad = MobileAd.getAdForId(id);
    if (ad == null) {
      result.error("ad_not_loaded", "show failed, the specified ad was not loaded id=" + id, null);
      return;
    }
    final String anchorOffset = call.argument("anchorOffset");
    final String anchorType = call.argument("anchorType");
    if (anchorOffset != null) {
      ad.anchorOffset = Double.parseDouble(anchorOffset);
    }
    if (anchorType != null) {
      ad.anchorType = "bottom".equals(anchorType) ? Gravity.BOTTOM : Gravity.TOP;
    }

    ad.show();
    result.success(Boolean.TRUE);
  }

  private void callIsAdLoaded(Integer id, Result result) {
    MobileAd ad = MobileAd.getAdForId(id);
    if (ad == null) {
      result.error("no_ad_for_id", "isAdLoaded failed, no add exists for id=" + id, null);
      return;
    }
    result.success(ad.status == MobileAd.Status.LOADED ? Boolean.TRUE : Boolean.FALSE);
  }

  private void callShowRewardedVideoAd(Result result) {
    if (rewardedWrapper.getStatus() == RewardedVideoAdWrapper.Status.LOADED) {
      rewardedWrapper.show();
      result.success(Boolean.TRUE);
    } else {
      result.error("ad_not_loaded", "show failed for rewarded video, no ad was loaded", null);
    }
  }

  private void callDisposeAd(Integer id, Result result) {
    MobileAd ad = MobileAd.getAdForId(id);
    if (ad == null) {
      result.error("no_ad_for_id", "dispose failed, no add exists for id=" + id, null);
      return;
    }

    ad.dispose();
    result.success(Boolean.TRUE);
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {

    Activity activity = registrar.activity();
    if (activity == null) {
      result.error("no_activity", "firebase_admob plugin requires a foreground activity", null);
      return;
    }

    Integer id = call.argument("id");

    switch (call.method) {
      case "initialize":
        callInitialize(call, result);
        break;
      case "loadBannerAd":
        callLoadBannerAd(id, activity, call, result);
        break;
      case "loadInterstitialAd":
        callLoadInterstitialAd(MobileAd.createInterstitial(id, activity, channel), call, result);
        break;
      case "loadRewardedVideoAd":
        callLoadRewardedVideoAd(call, result);
        break;
      case "showAd":
        callShowAd(id, call, result);
        break;
      case "showRewardedVideoAd":
        callShowRewardedVideoAd(result);
        break;
      case "disposeAd":
        callDisposeAd(id, result);
        break;
      case "isAdLoaded":
        callIsAdLoaded(id, result);
        break;
      default:
        result.notImplemented();
    }
  }
}

File: packages/firebase_admob/ios/Classes/FirebaseAdMobPlugin.m
// Copyright 2017, the Flutter project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

#import "FirebaseAdMobPlugin.h"

#import <UIKit/UIKit.h>

#import "FLTMobileAd.h"
#import "FLTRewardedVideoAdWrapper.h"
#import "Firebase/Firebase.h"

@interface NSError (FlutterError)
@property(readonly, nonatomic) FlutterError *flutterError;
@end

@implementation NSError (FlutterError)
- (FlutterError *)flutterError {
  return [FlutterError errorWithCode:[NSString stringWithFormat:@"%ld", (long)self.code]
                             message:self.domain
                             details:self.localizedDescription];
}
@end

@interface FLTFirebaseAdMobPlugin ()
@property(nonatomic, retain) FlutterMethodChannel *channel;
@property(nonatomic, strong) FLTRewardedVideoAdWrapper *rewardedWrapper;
@end

@implementation FLTFirebaseAdMobPlugin

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
  FLTFirebaseAdMobPlugin *instance = [[FLTFirebaseAdMobPlugin alloc] init];
  instance.channel =
      [FlutterMethodChannel methodChannelWithName:@"plugins.flutter.io/firebase_admob"
                                  binaryMessenger:[registrar messenger]];
  [registrar addMethodCallDelegate:instance channel:instance.channel];
  instance.rewardedWrapper = [[FLTRewardedVideoAdWrapper alloc] initWithChannel:instance.channel];
}

- (instancetype)init {
  self = [super init];
  if (self && ![FIRApp defaultApp]) {
    FLTLogWarning(@"[FIRApp configure]");
    [FIRApp configure];
  }
  return self;
}

- (void)dealloc {
  [self.channel setMethodCallHandler:nil];
  self.channel = nil;
}

- (void)callInitialize:(FlutterMethodCall *)call result:(FlutterResult)result {
  NSString *appId = (NSString *)call.arguments[@"appId"];
  if (appId == nil || [appId length] == 0) {
    result([FlutterError errorWithCode:@"no_app_id"
                               message:@"a non-empty AdMob appId was not provided"
                               details:nil]);
    return;
  }
  [FLTMobileAd configureWithAppId:appId];
  result([NSNumber numberWithBool:YES]);
}

- (void)callLoadBannerAdWithId:(NSNumber *)id
                       channel:(FlutterMethodChannel *)channel
                          call:(FlutterMethodCall *)call
                        result:(FlutterResult)result {
  NSString *adUnitId = (NSString *)call.arguments[@"adUnitId"];
  if (adUnitId == nil || [adUnitId length] == 0) {
    NSString *message =
        [NSString stringWithFormat:@"a null or empty adUnitId was provided for %@", id];
    result([FlutterError errorWithCode:@"no_adunit_id" message:message details:nil]);
    return;
  }

  NSNumber *widthArg = (NSNumber *)call.arguments[@"width"];
  NSNumber *heightArg = (NSNumber *)call.arguments[@"height"];

  if (widthArg == nil || heightArg == nil) {
    NSString *message =
        [NSString stringWithFormat:@"a null height or width was provided for banner id=%@", id];
    result([FlutterError errorWithCode:@"invalid_adsize" message:message details:nil]);
    return;
  }

  NSString *adSizeTypeArg = (NSString *)call.arguments[@"adSizeType"];
  FLTLogWarning(@"Size Type: %@", adSizeTypeArg);
  if (adSizeTypeArg == nil || (![adSizeTypeArg isEqualToString:@"AdSizeType.SmartBanner"] &&
                               ![adSizeTypeArg isEqualToString:@"AdSizeType.WidthAndHeight"])) {
    NSString *message = [NSString
        stringWithFormat:@"a null or invalid ad size type was provided for banner id=%@", id];
    result([FlutterError errorWithCode:@"invalid_adsizetype" message:message details:nil]);
    return;
  }

  int width = [widthArg intValue];
  int height = [heightArg intValue];

  if ([adSizeTypeArg isEqualToString:@"AdSizeType.WidthAndHeight"] && (width <= 0 || height <= 0)) {
    NSString *message =
        [NSString stringWithFormat:@"an invalid AdSize (%d, %d) was provided for banner id=%@",
                                   width, height, id];
    result([FlutterError errorWithCode:@"invalid_adsize" message:message details:nil]);
    return;
  }

  GADAdSize adSize;
  if ([adSizeTypeArg isEqualToString:@"AdSizeType.SmartBanner"]) {
    if (UIDeviceOrientationIsPortrait([UIDevice currentDevice].orientation)) {
      adSize = kGADAdSizeSmartBannerPortrait;
    } else {
      adSize = kGADAdSizeSmartBannerLandscape;
    }
  } else {
    adSize = GADAdSizeFromCGSize(CGSizeMake(width, height));
  }

  FLTBannerAd *banner = [FLTBannerAd withId:id adSize:adSize channel:self.channel];

  if (banner.status != CREATED) {
    if (banner.status == FAILED) {
      NSString *message = [NSString stringWithFormat:@"cannot reload a failed ad=%@", banner];
      result([FlutterError errorWithCode:@"load_failed_ad" message:message details:nil]);
    } else {
      result([NSNumber numberWithBool:YES]);  // The ad was already loaded.
    }
  }

  NSDictionary *targetingInfo = (NSDictionary *)call.arguments[@"targetingInfo"];
  [banner loadWithAdUnitId:adUnitId targetingInfo:targetingInfo];
  result([NSNumber numberWithBool:YES]);
}

- (void)callLoadInterstitialAd:(FLTMobileAd *)ad
                          call:(FlutterMethodCall *)call
                        result:(FlutterResult)result {
  if (ad.status != CREATED) {
    if (ad.status == FAILED) {
      NSString *message = [NSString stringWithFormat:@"cannot reload a failed ad=%@", ad];
      result([FlutterError errorWithCode:@"load_failed_ad" message:message details:nil]);
    } else {
      result([NSNumber numberWithBool:YES]);  // The ad was already loaded.
    }
  }

  NSString *adUnitId = (NSString *)call.arguments[@"adUnitId"];
  if (adUnitId == nil || [adUnitId length] == 0) {
    NSString *message =
        [NSString stringWithFormat:@"a null or emtpy adUnitId was provided for %@", ad];
    result([FlutterError errorWithCode:@"no_adunit_id" message:message details:nil]);
    return;
  }

  NSDictionary *targetingInfo = (NSDictionary *)call.arguments[@"targetingInfo"];
  [ad loadWithAdUnitId:adUnitId targetingInfo:targetingInfo];
  result([NSNumber numberWithBool:YES]);
}

- (void)callLoadRewardedVideoAd:(FlutterMethodCall *)call result:(FlutterResult)result {
  if (self.rewardedWrapper.status == FLTRewardedVideoAdStatusLoading ||
      self.rewardedWrapper.status == FLTRewardedVideoAdStatusLoaded) {
    result([NSNumber numberWithBool:YES]);  // The ad is loaded or about to be.
  }

  NSString *adUnitId = (NSString *)call.arguments[@"adUnitId"];
  if (adUnitId == nil || [adUnitId length] == 0) {
    result([FlutterError errorWithCode:@"no_ad_unit_id"
                               message:@"a non-empty adUnitId was not provided for rewarded video."
                               details:nil]);
    return;
  }

  NSDictionary *targetingInfo = (NSDictionary *)call.arguments[@"targetingInfo"];
  if (targetingInfo == nil) {
    result([FlutterError
        errorWithCode:@"no_targeting_info"
              message:@"a null targetingInfo object was provided for rewarded video."
              details:nil]);
    return;
  }

  [self.rewardedWrapper loadWithAdUnitId:adUnitId targetingInfo:targetingInfo];
  result([NSNumber numberWithBool:YES]);
}

- (void)callShowAd:(NSNumber *)mobileAdId
              call:(FlutterMethodCall *)call
            result:(FlutterResult)result {
  FLTMobileAd *ad = [FLTMobileAd getAdForId:mobileAdId];
  if (ad == nil) {
    NSString *message =
        [NSString stringWithFormat:@"show failed, the specified ad was not loaded id=%d",
                                   mobileAdId.intValue];
    result([FlutterError errorWithCode:@"ad_not_loaded" message:message details:nil]);
  }

  double offset = 0.0;
  int type = 0;
  if (call.arguments[@"anchorOffset"] != nil) {
    offset = [call.arguments[@"anchorOffset"] doubleValue];
  }
  if (call.arguments[@"anchorType"] != nil) {
    type = [call.arguments[@"anchorType"] isEqualToString:@"bottom"] ? 0 : 1;
  }

  [ad showAtOffset:offset fromAnchor:type];
  result([NSNumber numberWithBool:YES]);
}

- (void)callIsAdLoaded:(NSNumber *)mobileAdId
                  call:(FlutterMethodCall *)call
                result:(FlutterResult)result {
  FLTMobileAd *ad = [FLTMobileAd getAdForId:mobileAdId];
  if (ad == nil) {
    NSString *message = [NSString
        stringWithFormat:@"isAdLoaded failed, no ad exists for id=%d", mobileAdId.intValue];
    result([FlutterError errorWithCode:@"no_ad_for_id" message:message details:nil]);
    return;
  }
  if (ad.status == LOADED) {
    result([NSNumber numberWithBool:YES]);
  } else {
    result([NSNumber numberWithBool:NO]);
  }
}

- (void)callShowRewardedVideoAd:(FlutterMethodCall *)call result:(FlutterResult)result {
  if (self.rewardedWrapper.status != FLTRewardedVideoAdStatusLoaded) {
    result([FlutterError errorWithCode:@"ad_not_loaded"
                               message:@"show failed for rewarded video, no ad was loaded"
                               details:nil]);
    return;
  }

  [self.rewardedWrapper show];
  result([NSNumber numberWithBool:YES]);
}

- (void)callDisposeAd:(NSNumber *)mobileAdId
                 call:(FlutterMethodCall *)call
               result:(FlutterResult)result {
  FLTMobileAd *ad = [FLTMobileAd getAdForId:mobileAdId];
  if (ad == nil) {
    NSString *message =
        [NSString stringWithFormat:@"dispose failed, no ad exists for id=%d", mobileAdId.intValue];
    result([FlutterError errorWithCode:@"no_ad_for_id" message:message details:nil]);
  }

  [ad dispose];
  result([NSNumber numberWithBool:YES]);
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
  if ([call.method isEqualToString:@"initialize"]) {
    [self callInitialize:call result:result];
    return;
  }

  if ([call.method isEqualToString:@"loadRewardedVideoAd"]) {
    [self callLoadRewardedVideoAd:call result:result];
    return;
  }

  if ([call.method isEqualToString:@"showRewardedVideoAd"]) {
    [self callShowRewardedVideoAd:call result:result];
    return;
  }

  NSNumber *mobileAdId = (NSNumber *)call.arguments[@"id"];
  if (mobileAdId == nil) {
    NSString *message = @"FirebaseAdMobPlugin method calls for banners and "
                        @"interstitials must specify an "
                        @"integer mobile ad id";
    result([FlutterError errorWithCode:@"no_id" message:message details:nil]);
    return;
  }

  if ([call.method isEqualToString:@"loadBannerAd"]) {
    [self callLoadBannerAdWithId:mobileAdId channel:self.channel call:call result:result];
  } else if ([call.method isEqualToString:@"loadInterstitialAd"]) {
    [self callLoadInterstitialAd:[FLTInterstitialAd withId:mobileAdId channel:self.channel]
                            call:call
                          result:result];
  } else if ([call.method isEqualToString:@"showAd"]) {
    [self callShowAd:mobileAdId call:call result:result];
  } else if ([call.method isEqualToString:@"isAdLoaded"]) {
    [self callIsAdLoaded:mobileAdId call:call result:result];
  } else if ([call.method isEqualToString:@"disposeAd"]) {
    [self callDisposeAd:mobileAdId call:call result:result];
  } else {
    result(FlutterMethodNotImplemented);
  }
}

@end
